Optimize copyforward scheme for abort case and hybrid case

LinHu2016 opened the pull request:
In pull request #2855 , we introduced -XXgc:tarokEnableCopyForwardMarkCompactHybrid option to enable copyforward Hybrid mode to minimize MarkCompact PGC due to jniCritical in eden regions( it could reduce  more than 40% gc pause for the case). 
In order to measure the performance of  copyforwardHybrid in various cases, use testing option -XXgc:fvtest_forceCopyForwardMarkCompactHybridRatio can random pick from 1%-100% collectionSet regions to set markCompact, for extreme case -XXgc:fvtest_forceCopyForwardMarkCompactHybridRatio=100%, we expects similar performance as jvm is running with -XXgc:tarokPGCOnlyMarkCompact option.
But we did see performance gap between CopyForwardMarkCompactHybridRatio=100% and PGCOnlyMarkCompact(Hybrid shows 79% more GMP cycles, 25% more PGC pause, 5% throughput drop).
The performance gap are majorly caused by the below 3 issues.
1, mark in copyforward (include abort case) does not aware  leaf objects(there is no reference in leaf object, don't need to add them in scan stack) during scan.  
2, underestimate the remaining after PGC due to overestimate the required survivor regions
3, missed to adjust "live" bytes for mark only regions (both abort and hybrid case), it would cause to pick wrong RateOfReturnCollectionSet for the next collection(lower return rate).

This pull request tries to optimize code to avoid the 3 issues.

	Optimize copyforward scheme for abort case and hybrid case

	- Allow jniCritical eden regions	are part of nursery
	collectionSet in CopyForwardHybrid mode.
	- report the numbers of regions in collection set (for eden,
	nonEden, evacuated, marked) in gc-op-copypforward of verbose gc.
	- Not push leaf Objects into work stack for abort case and
	 CopyforwardHybrid mode.
	- create inline method iterateAndCopyforwardSlotReference()
	for avoiding duplicated logic.
	- Perform Atomic Sweep on nonEvacurated regions before compacting
	for abort case and CopyforwardHybrid mode in order to maintain
	accurate projectedLiveByte for the regions.
	- Correct estimatePartialGCsRemaining if
	_extensions->fvtest_forceCopyForwardHybridRatio is set
	(testing purpose) to avoid underestimating the remaining.
    - avoid any MarkCompactPGC in Hybrid mode to reduce longer gc pause
      1, eliminate MarkCompactPGCs caused by jniCrital(set any
      jniCritial Eden regions to mark only).
      2, eliminate MarkCompactPGCs caused by disabling copyforward
      due to pervious abort during GMP(set number of Eden regions to
      mark only if estimating free regions are not enough for survivor).

Signed-off-by: Lin Hu <linhu@ca.ibm.com>

LinHu2016 commented:
this PR would help to reduce GMP cycles and especially maximum gc pause.

SPECjbb2015 result  | disable Hybrid mode | enable Hybrid mode | Optimazed Hybrid mode
  ------------ | ------------- | ------------- | ------------- 
GC Mode | balanced | balanced | balanced
Global collections - Mean garbage collection pause (ms) | 813 | 719 | 166
Global collections - Mean interval between collections (ms) | 86092 | 13518 | 1895881
Global collections - Number of collections | 1 | 1 | 1
Global mark phase collections - Mean garbage collection pause (ms) | 55.3 | 49.8 | 38.9
Global mark phase collections - Mean interval between collections (ms) | 34576 | 46033 | 40997
Global mark phase collections - Number of collections | 42 | 57 | 47
Global mark phase cycles - Mean collections per cycle | 3.08 | 3.6 | 4.27
Global mark phase cycles - Number of cycles | 13 | 15 | 11
Mean heap unusable due to fragmentation (MB) | 656 | 632 | 655
Number of collections triggered by allocation failure | 0 | 0 | 0
Partial collections - Mean garbage collection pause (ms) | 147 | 141 | 141
Partial collections - Mean interval between collections (ms) | 2624| 2608 | 2625
Partial collections - Number of collections | 706 | 710 | 699
Proportion of time spent in garbage collection pauses (%) | 6.04 | 5.81 | 5.67
Proportion of time spent unpaused (%) | 93.96 | 94.19 | 94.33
Rate of garbage collection (MB/minutes) | 46893 | 47150 | 46872
Mean Pause time (ms) | 143 | 135 | 135
Maximum Pause time (ms)| 813 | 719 | 296
Total Pause time (ms)| 107144 | 103657 | 100549

LinHu2016 commented:
SPECjbb2015 Result | tarokPGCOnlyMarkCompact | fvtest_forceCopyForwardHybridMarkCompactRatio=100 | Optimazed fvtest_forceCopyForwardHybridMarkCompactRatio=100
-------------- | ---------------- | ---------------- | --------------------
Forced collection count | 0 | 0 | 0
GC Mode | balanced | balanced | balanced
Global collections - Mean garbage collection pause (ms) | 0 | 0 | 0
Global collections - Number of collections | 0 | 0 | 0
Global mark phase collections - Mean garbage collection pause (ms) | 128 | 150 | 135
Global mark phase collections - Mean interval between collections (ms) | 29475 | 21988 | 30196
Global mark phase collections - Number of collections | 56 | 78 | 54
Global mark phase cycles - Mean collections per cycle | 2 | 2 | 2
Global mark phase cycles - Number of cycles | 28 | 39 | 27
Mean heap unusable due to fragmentation (MB) | 560 | 564 | 570
Number of collections triggered by allocation failure | 0 | 0 | 0
Partial collections - Mean garbage collection pause (ms) | 416 | 463 | 426
Partial collections - Mean interval between collections (ms) | 2640 | 2643 | 2643
Partial collections - Number of collections | 606 | 606 | 605
Proportion of time spent in garbage collection pauses (%) | 16.53 | 18.5 | 16.89
Proportion of time spent unpaused (%) | 83.47 | 81.5 | 83.11
Rate of garbage collection (MB/minutes) | 46628 | 46572 | 46569
Mean Pause time (ms) | 393 | 427 | 403
Maximum Pause time (ms) | 678 | 584 | 651
Total Pause time (ms) | 260324 | 292171 | 265864

LinHu2016 commented:
Hi @amicic, the pull request is ready for review, Thanks,

LinHu2016 commented:
the code has passed personal build, current Travis CI build failure is only related with sourceforge download issue.

amicic commented on runtime/gc_verbose_handler_vlhgc/VerboseHandlerOutputVLHGC.cpp:
```diff
@@ -427,7 +427,8 @@ MM_VerboseHandlerOutputVLHGC::handleCopyForwardEnd(J9HookInterface** hook, UDATA
 	}
 	if (0 != copyForwardStats->_nonEvacuateRegionCount) {
-		writer->formatAndOutput(env, 1, "<warning details=\"running under hybrid mode, JNI Critical region count=%zu\" />", copyForwardStats->_nonEvacuateRegionCount);
+		writer->formatAndOutput(env, 1, "<warning details=\"running under hybrid mode, Non Evacurated region count=%zu, scanObjects=\"%zu\" markobjects=\"%zu\" scanBytes=\"%zu\" />",
```
I would not report this as warning. Besides, we already report memory-trace stanza just a few lines above - could we not piggy back on it (for scan/mark objects)?
But yes, it would be nice to report additionally, how many regions were not subject to evacuation (in a separate line).

amicic replied:
there is spelling error here (and at several other spots - please search): Evacurate

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -1316,9 +1316,34 @@ MM_IncrementalGenerationalGC::partialGarbageCollectUsingCopyForward(MM_Environme
+	if ((0 != extensions->fvtest_forceCopyForwardHybridRatio) && (100 >= extensions->fvtest_forceCopyForwardHybridRatio)) {
+		estimatedReguiredSurvivorRegions = estimatedReguiredSurvivorRegions * (100 - extensions->fvtest_forceCopyForwardHybridRatio) / 100;
+	}
+	j9tty_printf(PORTLIB, "partialGarbageCollectUsingCopyForward freeRegions=%zu, estimatedReguiredSurvivorRegions=%f\n", freeRegions, estimatedReguiredSurvivorRegions);
```
remnant print here

LinHu2016 replied:
the printfs are for my local debug, I will remove them.

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -1316,9 +1316,34 @@ MM_IncrementalGenerationalGC::partialGarbageCollectUsingCopyForward(MM_Environme
+
+	if ((_extensions->tarokEnableCopyForwardHybrid || (0 != _extensions->fvtest_forceCopyForwardHybridRatio)) && (_schedulingDelegate.isPGCAbortDuringGMP() || _schedulingDelegate.isFirstPGCAfterGMP()) && (estimatedReguiredSurvivorRegions > freeRegions)) {
+		double edenSurvivorRate = _schedulingDelegate.getAvgEdenSurvivalRateCopyForward(env);
+		j9tty_printf(PORTLIB, "edenSurvivorRate=%f\n", edenSurvivorRate);
```
print

amicic commented on runtime/gc_vlhgc/IncrementalGenerationalGC.cpp:
```diff
@@ -1316,9 +1316,34 @@ MM_IncrementalGenerationalGC::partialGarbageCollectUsingCopyForward(MM_Environme
+			regionCountRequiredMarkOnly = _schedulingDelegate.getCurrentEdenSizeInRegions(env);
+		}
+		/* set the number of the selected eden regions to nonEvacurated region to avoid potential abort case */
+		j9tty_printf(PORTLIB, "regionCountRequiredMarkOnly=%zu\n", regionCountRequiredMarkOnly);
```
print

LinHu2016 commented:
update as @amicic suggestions.
removed - Allow jniCritical regions(non eden regions)	are part of RateOfReturn CollectionSet in CopyForwardHybrid mode
changed - report the numbers of regions in collection set (for eden, nonEden, evacuated, marked) in gc-op-copypforward of verbose gc
created - new inline method iteratorAndCopyforwardSlotReference() for avoiding duplicated logic.

LinHu2016 commented:
@amicic update is done.

amicic commented on runtime/gc_vlhgc/CopyForwardScheme.cpp:
```diff
@@ -2382,6 +2396,70 @@ MM_CopyForwardScheme::scanOwnableSynchronizerObjectSlots(MM_EnvironmentVLHGC *en
+ *  to avoid to push leaf object to work stack in case the reference need to be marked instead of copied.
+ */
+MMINLINE bool
+MM_CopyForwardScheme::iteratorAndCopyforwardSlotReference(MM_EnvironmentVLHGC *env, MM_AllocationContextTarok *reservingContext, J9Object *objectPtr) {
```
rename it to iterateAndCopyForwardSlots (use verb, rather than noun)

LinHu2016 replied:
done

amicic commented on runtime/gc_vlhgc/SchedulingDelegate.cpp:
```diff
@@ -598,6 +599,10 @@ MM_SchedulingDelegate::estimatePartialGCsRemaining(MM_EnvironmentVLHGC *env) con
 			/* Calculate the number of regions that we need for copy forward destination */
 			double survivorRegions = _averageSurvivorSetRegionCount;
+			/* if _extensions->fvtest_forceCopyForwardHybridRatio is set(testing purpose), correct required survivor region count to avoid underestimating the remaining. */
+			if (0 != _extensions->fvtest_forceCopyForwardHybridRatio && 100 >= _extensions->fvtest_forceCopyForwardHybridRatio) {
```
add extra brackets for both conditions

LinHu2016 replied:
done

amicic commented:
Jenkins test sanity xLinux jdk8

amicic commented:
Jenkins test sanity xLinux jdk8

amicic commented:
Jenkins test sanity xLinux jdk8
